sdk/core/core-util/test/public/reshape.spec.ts

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { describe, it, assert, expect } from "vitest";
+import { reshape } from "../../src/reshape";
+
+describe("reshape", function () {
+  it("renames a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", "c");
+    const expected = { c: 1, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("maps a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", String);
+    const expected = { a: "1", b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renames and maps a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", "c", String);
+    const expected = { c: "1", b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renames a nested property", function () {
+    const input = { a: { c: { e: 4 }, d: 3 }, b: 2 };
+    const actual = reshape(input, "a.c.e", "f");
+    const expected = { a: { c: { f: 4 }, d: 3 }, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("can map elements of an array", function () {
+    const input = { a: { b: [1, 2, 3] } };
+    const actual = reshape(input, "a.b[]", String);
+    const expected = { a: { b: ["1", "2", "3"] } };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renaming array elements is an error", function () {
+    const input = { a: { b: [1, 2, 3] } };
+    expect(() => reshape(input, "a.b[]", "oops")).toThrowError("cannot rename array indices");
+  });
+
+  it("can rename properties of elements of an array", function () {
+    const input = {
+      a: {
+        b: [
+          { c: 1, d: 2 },
+          { c: 3, d: 4 },
+        ],
+      },
+    };
+    const actual = reshape(input, "a.b[]c", "e");
+    const expected = {
+      a: {
+        b: [
+          { e: 1, d: 2 },
+          { e: 3, d: 4 },
+        ],
+      },
+    };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("selecting a property that doesn't exist does nothing", function () {
```

Comment:
just curious if we should return undefined if the selected property doesn't exist? put it another way, do we trust the selector or do we trust the input data ?

Reply:
The selector doesn't really determine what gets returned, it just controls what gets renamed / mapped to a new value. 

If you try to select something that isn't there, there's nothing to perform the operation on. This helps with the case of unions where we don't actually have to know which variant we are processing if we know we only want to rename certain properties if they exist.

Consider the OpenAI case where it looks like `Array<{url: string} | {b64_json: string}>`, if we are selecting something like `prop[]b64_json` then we can do the rename if that property exists without worrying that we will mutate an object that only has `url`